---
title: "R homework"
output: html_document
---

##Murphy Review Comments:

I made several comments below referencing specific sections of your code.
(My comments begin with "#@")

Overall, you should refrain from including commands to run your files every time you create them, since
this consumes a LOT of time and memory! (And it actually crashed my R Studio several times and
required me to reinstall the program completely.) I've placed comment tags before these commands
in your code below to prevent all those files from running.




```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Part I: Data inspection and processing

1 Get the origianl files


```{r}
library(tidyverse)
genotype <- read_tsv("https://raw.githubusercontent.com/EEOB-BioData/BCB546X-Fall2018/master/assignments/UNIX_Assignment/fang_et_al_genotypes.txt")
#genotype
snp <- read_tsv("https://raw.githubusercontent.com/EEOB-BioData/BCB546X-Fall2018/master/assignments/UNIX_Assignment/snp_position.txt")
#snp
```

2 inspect both files

```{r}
is.data.frame(genotype)
is.data.frame(snp)
ncol(genotype)
nrow(genotype)
ncol(snp)
nrow(snp)
```

3 original genotype file processing (extracting maize and teosinte data,respectly and transpose the extrated data to new data.frames)

```{r}
library(dplyr)
mg <- filter (genotype, Group %in% c("ZMMIL","ZMMMR","ZMMLR")) # extracting maize data
teog <- filter (genotype, Group %in% c ("ZMPBA","ZMPIL","ZMPJA")) # extracting teosinte data
#mg # check the extracted data
transposed_mg <- t(mg)%>% as.data.frame() # transpose maize genotype data so columns become rows and keep them as data.frame 
transposed_teog <- t(teog)%>% as.data.frame() # transpose teosinte genotype data so columns become rows and keep them as data.frame 
#transposed_mg # check the transposed data
#transposed_teog # check the transposed data
s_mg <- slice(transposed_mg, 4:986) # get rid of first 3 rows 
s_teog <- slice(transposed_teog,4:986) # get rid of first 3 rows

#@ Your transposed files still contain the various SNP IDs as row names, but when you slice these
#@ files you end up losing the row names in your s_mg and s_teog files.
```
4 snp file processing
```{r}
#snp
new_snp <- select(snp, SNP_ID, Chromosome, Position) # select col 1,3,4 to a new data.frame named new_snp
#new_snp # check the new data.frame

```

5 combine the data from processed genotype and snp files

```{r}
c_sm <- bind_cols(new_snp,s_mg) # combine data from SNP file (col 1,3,4) and data from maize
#c_sm
c_steog <- bind_cols(new_snp,s_teog) # combine data from SNP file (col 1,3,4) and data from teosinte
#c_steog

#@ This approach assumes that the rows from your `s_mg` and `s_teog` files happen to align perfectly
#@ with the SNP IDs listed in the first column of `new_snp`.  Though that may be the case in this instance,
#@ overall it is an extremely fragile approach, and one for which the resulting errors would not be
#@ visible in your downstream analyses!!
```
6 arrange the combined files with SNPs ordered based on increasing position values and then subset them based on chromosome number

```{r}
c_sm <- filter(c_sm,Chromosome %in% c("1","2","4","5","6","7","8","9","10","3")&Position !="multiple")# get rid of rows with "unknown" and "multiple" at col2 and col3
c_steog <- filter(c_steog,Chromosome %in% c("1","2","4","5","6","7","8","9","10","3")&Position !="multiple")
c_sm <-transform(c_sm, Chromosome =as.numeric(Chromosome),Position =as.numeric(Position))# transform variables (col2,col3) from character into numeric
c_steog <-transform(c_steog, Chromosome =as.numeric(Chromosome),Position =as.numeric(Position))
c_sm <- arrange(c_sm,Chromosome, Position)
c_steog<- arrange(c_steog,Chromosome,Position) 


uniqm <- unique(unlist(c_sm$Chromosome))
#uniqm
for (i in 1:length(uniqm)){
  filename <- paste("maize_chr",i,".txt", sep="")
  print(filename)
  c_sm %>% filter(Chromosome==uniqm[i]) %>%
    write.csv(filename)
} # extracting different rows based on Chromosome number

uniqt <- unique(unlist(c_steog$Chromosome))
for (i in 1:length(uniqt)){
  filename <- paste("teo_chr",i,".txt", sep="")
  print(filename)
  c_sm %>% filter(Chromosome==uniqm[i]) %>%
    write.csv(filename)
}

#@ Excellent use of `for` loops to create your chromosome files!
```

7 replace "?" with "-" and then arrange and subset files based on chromosome and decreasing order of position

```{r}
c_sm %>% mutate_if(is.factor, as.character) -> c_sm # convert factor into character before using gsub 
c_steog %>% mutate_if(is.factor, as.character) -> c_steog
c_sm[-c(1,2,3)] <- lapply(c_sm[-c(1,2,3)],gsub,pattern="?",replacement="-",fixed=T)
c_steog[-c(1,2,3)] <-lapply(c_steog[-c(1,2,3)],gsub,pattern="?",replacement="-",fixed=T)
rd_c_sm <- arrange(c_sm, Chromosome,desc(Position))
rd_c_steog <- arrange(c_sm, Chromosome,desc(Position))



for (i in 1:length(uniqm)){
  filename <- paste("d_maize_chr",i,".txt", sep="")
  print(filename)
  rd_c_sm %>% filter(Chromosome==uniqm[i]) %>%
    write.csv(filename)
}
for (i in 1:length(uniqt)){
  filename <- paste("d_teo_chr",i,".txt", sep="")
  print(filename)
  rd_c_steog %>% filter(Chromosome==uniqm[i]) %>%
    write.csv(filename)
}

```

## PartII: Data visualization

1 data reshaping

```{r}
install.packages("reshape2")
library(reshape2)

#@ Also need `ggplot2` for this section:

library(ggplot2)

tidy_m <- melt(c_sm, id.vars =c("SNP_ID", "Chromosome", "Position"),variable.name = "corn",value.name = "base.composition")

tidy_teo <- melt(c_steog, id.vars =c("SNP_ID", "Chromosome", "Position"),variable.name = "corn",value.name = "base.composition")


```

2 creat graph based on my own interested feature
```{r}
ggplot(data=tidy_m)+geom_point(mapping=aes(x=Chromosome,y=Position),alpha=0.01)+scale_x_discrete(limits=c(1,2,3,4,5,6,7,8,9,10))# SNPs distribution along each chromosome in Maize
ggsave("SNPs_distribution_maize_chromosome.png")
ggplot(data=tidy_teo)+geom_point(mapping=aes(x=Chromosome,y=Position),alpha=0.01)+scale_x_discrete(limits=c(1,2,3,4,5,6,7,8,9,10))#SNPs distribution along each chromosome in teosinte
ggsave("SNPs_distribution_teosinte_chromosome.png")
```

3 creating graphs for 1)SNPs per chromosome;2)SNPs per Group;3) percentage of homozygosity,hetetozygousity and missing data

```{r}
ggplot(data=new_snp)+geom_bar(mapping=aes(x=Chromosome))+scale_x_discrete(limits=c(1,2,3,4,5,6,7,8,9,10))
ggsave("SNPs_per_Chromosome.png")

#@ You forgot to remove the "multiple/unknown" entries from `new_snp` before graphing (hence the error
#@ printed with the graph).

tidy_genotype<-  melt(genotype, id.vars =c("Sample_ID", "JG_OTU", "Group"),variable.name = "SNPs_ID",value.name = "base.composition")

tidy_genotype$Homozygous <- TRUE
tidy_genotype$Homozygous[tidy_genotype$base.composition=="?/?"]<-NA
tidy_genotype$Homozygous[substr(tidy_genotype$base.composition,1,1)!=substr(tidy_genotype$base.composition,3,3)] <- FALSE

#@ Very clever way to identify heterozygous sites!

tidy_genotype <- arrange(tidy_genotype, Group)
ftg <- filter (tidy_genotype, base.composition!="?/?")
s_ftg <- ftg %>% group_by(Group, SNPs_ID, add = TRUE) %>% distinct(base.composition) %>% summarize(n())
s_ftg1 <- s_ftg %>% mutate(snp_present = (`n()` > 1))%>% group_by(Group) %>% summarize(sum(snp_present))
ggplot(s_ftg1,aes(Group, `sum(snp_present)`)) + geom_col()+theme(axis.text.x = element_text(angle = 90,hjust = 1))
ggsave("SNPs_per_Group.png")


temp.plot <- ggplot(data = tidy_genotype)+ geom_bar(mapping = aes(x = Group, color = Homozygous, fill = Homozygous), position= "fill")+theme(axis.text.x = element_text(angle = 90,hjust = 1))
ggsave ("status_position.png")



```
